### DISCUSSION TITLE
Excessive throat length and flow rate of each pore

### DISCUSSION THREAD

**Discussion by Devin-one** [PMEAL/OpenPNM#2607] [Q&A]
Hello Professor. The network spacing I set is 1, but the throat length generated is 33.9, and the pore and throat diameters are normally 0.08 to 0. 12. I don't know what went wrong. By the way, if I want to calculate the flow rate of each pore, can I use `sf. A * sf.x `to calculate. Because after I calculated this way, I found that the order of magnitude of pore flow is very small. Here is my code. Thank you in advance for your reply.
```
import numpy as np
import matplotlib.pyplot as plt
import openpnm as op
import scipy.stats
op.visualization.set_mpl_style()
np.random.seed(0)
Lx, Ly = 200,80
pn = op.network.Cubic(shape=[Lx, Ly], spacing=1)
pn['pore.seed']= np.ones_like(pn.Ps, dtype=float)
pn['pore.seed']*= e
stats_obj = scipy.stats.weibull_min(c=1.99, scale=1.2, loc=0)
pn.add_model(propname='pore.diameter',
             model=op.models.geometry.pore_size.generic_distribution,
             seeds='pore.seed',
             func=stats_obj)
pn.add_model(propname='pore.volume',
              model=op.models.geometry.pore_volume.sphere)
pn.add_model(propname='pore.area',
              model=op.models.geometry.pore_cross_sectional_area.sphere)
pn.add_model(propname='throat.diameter',
              model=op.models.geometry.throat_size.from_neighbor_pores,
              mode='min')
pn.add_model(propname='throat.length',
              model=op.models.geometry.throat_length.spheres_and_cylinders)
pn.add_model(propname='throat.volume',
              model=op.models.geometry.throat_volume.cylinder)
pn.add_model(propname='throat.hydraulic_size_factors',
              model=op.models.geometry.hydraulic_size_factors.spheres_and_cylinders)
air = op.phase.Air(network=pn)
#air['pore.diffusivity'] = 1e-9
air['pore.viscosity'] = 0.000018314
air['pore.temperature']=290
air.add_model_collection(op.models.collections.phase.air)
air.add_model_collection(op.models.collections.physics.standard)
air.regenerate_models()
sf = op.algorithms.StokesFlow(network=pn, phase=air)
pn.set_label(label='in', pores=[0,1,2,3,4])
pn.set_label(label='out', pores=[75,76,77,78,79])
inlet  = pn.pores('in') 
outlet = pn.pores('out')
sf.set_value_BC(pores=inlet, values=43.0)
sf.set_value_BC(pores=outlet,values=0)
sf.run();
air.update({'pore.pressure':sf['pore.pressure']})
gi=sf.A
pi=sf.x
qi=gi*pi
```

**Comment by jgostick** [comment]
What is `e` on the line `pn['pore.seed']*= e`?

**Reply by Devin-one** [reply]
I'm sorry I didn't explain it clearly. To achieve the porosity I want, I create an array and assign it to the `pn['pore.seed']`. The following is the distribution map of `pn['pore.seed']` and other geometric characteristics.
![dfaa6f0b0023a90c97f69ea82146f68](https://user-images.githubusercontent.com/93040662/197697887-d9eee844-94e5-4cb1-92e0-db48687e83d9.jpg)
![8d7162dc8662256cdf34eda0dddaf82](https://user-images.githubusercontent.com/93040662/197701312-7ffb8ca6-9e05-481b-90a5-9413c9421811.jpg)

**Comment by Devin-one** [comment]
Hello Professor, I redefined the `pn['pore.seed']`.
```
R1, R2 = (pn['pore.diameter'][pn.conns]/2).T
L_total = np.sqrt(np.sum(np.diff(pn.coords[pn.conns], axis=1).squeeze()**2, axis=1))   
Lt = L_total - R1 - R2    
pn['throat.length']=Lt.
```
Now its value is within the normal range. However, there is still one problem unsolved, that is, the calculation of each pore flow. Can I use sf. A * sf.x to calculate. Because after I calculated this way, I found that the order of magnitude of pore flow is very small. Could you give me some advice？ Thank you in advance for your reply.

**Reply by jgostick** [reply]
I'm glad you've made some progress, but at this point things are so subtle that I'm not sure how to help. You basically need to check the numerical values of every thing that goes into the hydraulic conductance.

**Reply by Devin-one** [reply]
The purpose of calculating the flow rate of each pore is to obtain the velocity of each pore. Now I don't know how to calculate the flow of each pore. Could you help me? Thank you in advance for your reply.

**Reply by jgostick** [reply]
Pores have 0 flow, since what goes in also leaves, so net 0 flow.  Throats have a flow though.  So you can find that using the `rate` method on the algorithms, like `alg.rate(throats=pn.Ts, rate='single')` will return the rate of flow through each throat in the network.  Note that positive flow is from pore i to pore j, where i and j are given by `pn.conns`, with i in the first column and j in the second column.

**Comment by Devin-one** [comment]
Hello Professor,
          Thank you very much for your help. I try to use `q=sf.rate(pores=pn.Ts,mode='single')` to find throat rate, and then use `D=pn['throat.diameter']`    `A=3.14159*(D/2)**2`   ` v=q/A `to calculate throat velocity. Finally, generate the image with
` ax = op.visualization.plot_connections(pn,color='w') `   `ax = op.visualization.plot_coordinates(pn, ax=ax, color_by=v,c=v,cmap=cm)`. But the result I finally got was very unreasonable. Maybe I know too little about the throat. The  `q`  found above does not correspond to the `throat area`. So I'm sorry to bother you again.
Best wishes.

**Reply by jgostick** [reply]
`q=sf.rate(pores=pn.Ts,mode='single')` should be q=sf.rate(**throats**=pn.Ts,mode='single').  See if that helps?
